tests/python/multidevice/test_expert_parallel.py

```diff
@@ -0,0 +1,114 @@
+# SPDX-FileCopyrightText: Copyright (c) 2025-present NVIDIA CORPORATION & AFFILIATES.
+# All rights reserved.
+# SPDX-License-Identifier: BSD-3-Clause
+
+import pytest
+import torch
+import torch.distributed as dist
+
+
+# Partitions n into k integers whose sum is n.
+def random_partition(n: int, k: int) -> torch.Tensor:
+    offsets = torch.randint(0, n + 1, [k - 1]).tolist()
+    # The last partition ends at index n.
+    offsets.append(n)
+    offsets.sort()
+
+    sizes = []
+    prev = 0
+    for offset in offsets:
+        sizes.append(offset - prev)
+        prev = offset
+    assert sum(sizes) == n
+    assert len(sizes) == k
+    return torch.tensor(sizes, device="cuda")
+
+
+def list_to_matrix(l: list, rows: int, cols: int) -> list[list]:
+    return [l[row * cols : (row + 1) * cols] for row in range(rows)]
+
+
+def matrix_to_list(m: list[list]) -> list:
+    return [item for row in m for item in row]
+
+
+# This can be fused into one kernel.
+def rank_first_to_expert_first(
+    tokens: torch.Tensor, n_tokens_for_expert_from_rank: torch.Tensor
+) -> torch.Tensor:
+    d, num_experts_per_rank = n_tokens_for_expert_from_rank.size()
+    tokens_for_expert_from_rank = tokens.split(
+        n_tokens_for_expert_from_rank.flatten().tolist()
+    )
+    tokens_for_expert_from_rank = list_to_matrix(
+        tokens_for_expert_from_rank, d, num_experts_per_rank
+    )
+    # list(zip(*matrix)) transposes the matrix.
+    tokens_for_expert_from_rank = list(zip(*tokens_for_expert_from_rank))
+    tokens_for_expert_from_rank = matrix_to_list(tokens_for_expert_from_rank)
+    tokens = torch.cat(tokens_for_expert_from_rank, dim=0)
+    return tokens
+
+
+# This test serves as the reference implementation for the expert parallelism
+# dispatch logic. Each token is dispatched to the rank hosting the corresponding
+# expert, and tokens received by the same rank are sorted by expert ID.
+@pytest.mark.mpi
+def test_dispatch(setup_default_process_group):
+    rank = dist.get_rank()
+    # So RNG generates different numbers for each rank.
+    torch.manual_seed(rank)
+
+    d = dist.get_world_size()
+    num_experts_per_rank = 3
+    n_experts = d * num_experts_per_rank
+
+    # Input tokens are distributed among ranks, so the effective sequence length is `n_tokens * d`.
+    n_tokens = 12
+    # GPU 0: [n_tokens_for_expert_0, n_tokens_for_expert_1, ..., n_tokens_for_expert_5], sums to `n_tokens`
+    # GPU 1: [n_tokens_for_expert_0, n_tokens_for_expert_1, ..., n_tokens_for_expert_5], sums to `n_tokens`
+    n_tokens_for_expert = random_partition(n_tokens, n_experts)
```

Comment:
Trying to understand the logic around here. Is there any assumption about `d`? Looks like `n_tokens` must be `>= n_experts`, so is `d` assumed to be `<= 4`?

Reply:
> Looks like n_tokens must be >= n_experts

No, an expert can take zero tokens. I remember actually seeing some zeros when I print out n_tokens_for_expert.

Reply:
Yes. My next PR will make it more clear: https://github.com/NVIDIA/Fuser/pull/4939/files#diff-dca6ab3971a109c00c4301666c64ac3e6409be49292c77ed5ec8451a418e9bd7R89-R91